File: collect_app/src/androidTest/java/org/odk/collect/android/feature/maps/FormMapTest.java
package org.odk.collect.android.feature.maps;

import android.Manifest;
import android.app.Activity;
import android.app.Instrumentation;
import android.content.Intent;
import android.location.Location;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.rule.GrantPermissionRule;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.odk.collect.android.R;
import org.odk.collect.android.RecordedIntentsRule;
import org.odk.collect.android.activities.FormEntryActivity;
import org.odk.collect.android.support.CollectTestRule;
import org.odk.collect.android.support.CopyFormRule;
import org.odk.collect.android.support.ResetStateRule;
import org.odk.collect.android.support.pages.MainMenuPage;
import org.odk.collect.geo.GeoUtils;

import static androidx.test.espresso.intent.Intents.intending;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent;

public class FormMapTest {

    private static final String SINGLE_GEOPOINT_FORM = "single-geopoint.xml";
    private static final String NO_GEOPOINT_FORM = "basic.xml";

    public final CollectTestRule rule = new CollectTestRule();

    @Rule
    public RuleChain copyFormChain = RuleChain
            .outerRule(GrantPermissionRule.grant(Manifest.permission.ACCESS_COARSE_LOCATION))
            .around(new RecordedIntentsRule())
            .around(new ResetStateRule())
            .around(new CopyFormRule(SINGLE_GEOPOINT_FORM))
            .around(new CopyFormRule(NO_GEOPOINT_FORM))
            .around(rule);

    @Before
    public void stubGeopointIntent() {
        Location location = new Location("gps");
        location.setLatitude(125.1);
        location.setLongitude(10.1);
        location.setAltitude(5);

        Intent intent = new Intent();
        intent.putExtra(FormEntryActivity.ANSWER_KEY, GeoUtils.formatLocationResultString(location));
        Instrumentation.ActivityResult result = new Instrumentation.ActivityResult(Activity.RESULT_OK, intent);

        intending(hasComponent("org.odk.collect.geo.GeoPointActivity"))
                .respondWith(result);
    }

    @Test
    public void gettingBlankFormList_showsMapIcon_onlyForFormsWithGeometry() {
        new MainMenuPage()
                .clickFillBlankForm()
                .checkMapIconDisplayedForForm("Single geopoint")
                .checkMapIconNotDisplayedForForm("basic");
    }

    @Test
    public void clickingOnMapIcon_opensMapForForm() {
        new MainMenuPage()
                .clickFillBlankForm()
                .clickOnMapIconForForm("Single geopoint")
                .assertText("Single geopoint");
    }

    @Test
    public void fillingBlankForm_addsInstanceToMap() {
        String oneInstanceString = ApplicationProvider.getApplicationContext().getResources().getString(R.string.geometry_status, 1, 1);

        new MainMenuPage()
                .clickFillBlankForm()
                .clickOnMapIconForForm("Single geopoint")
                .clickFillBlankFormButton("Single geopoint")
                .inputText("Foo")
                .swipeToNextQuestion()
                .clickWidgetButton()
                .swipeToEndScreen()
                .clickSaveAndExitBackToMap()
                .assertText(oneInstanceString);
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/maps/FormMapTest.kt
package org.odk.collect.android.feature.maps

import android.app.Activity
import android.app.Instrumentation
import android.content.Context
import android.location.Location
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import androidx.test.espresso.intent.Intents.intending
import androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import org.odk.collect.android.R
import org.odk.collect.android.support.FakeClickableMapFragment
import org.odk.collect.android.support.TestDependencies
import org.odk.collect.android.support.rules.CollectTestRule
import org.odk.collect.android.support.rules.TestRuleChain
import org.odk.collect.androidtest.RecordedIntentsRule
import org.odk.collect.externalapp.ExternalAppUtils.getReturnIntent
import org.odk.collect.geo.GeoUtils
import org.odk.collect.maps.MapFragment
import org.odk.collect.maps.MapFragmentFactory
import org.odk.collect.settings.SettingsProvider

@RunWith(AndroidJUnit4::class)
class FormMapTest {

    private val mapFragment = FakeClickableMapFragment()
    private val testDependencies = object : TestDependencies() {
        override fun providesMapFragmentFactory(settingsProvider: SettingsProvider): MapFragmentFactory {
            return object : MapFragmentFactory {
                override fun createMapFragment(): MapFragment {
                    return mapFragment
                }
            }
        }
    }
    private val rule = CollectTestRule()

    @get:Rule
    var copyFormChain: RuleChain = TestRuleChain.chain(testDependencies)
        .around(RecordedIntentsRule())
        .around(rule)

    @Test
    fun gettingBlankFormList_showsMapIcon_onlyForFormsWithGeometry() {
        rule.startAtMainMenu()
            .copyForm(SINGLE_GEOPOINT_FORM)
            .copyForm(NO_GEOPOINT_FORM)
            .clickFillBlankForm()
            .checkMapIconDisplayedForForm("Single geopoint")
            .checkMapIconNotDisplayedForForm("basic")
    }

    @Test
    fun fillingBlankForm_addsInstanceToMap() {
        stubGeopointIntent()

        rule.startAtMainMenu()
            .copyForm(SINGLE_GEOPOINT_FORM)
            .copyForm(NO_GEOPOINT_FORM)
            .clickFillBlankForm()
            .clickOnMapIconForForm("Single geopoint")
            .clickFillBlankFormButton("Single geopoint")

            .inputText("Foo")
            .swipeToNextQuestion("Location")
            .clickWidgetButton()
            .swipeToEndScreen()
            .clickSaveAndExitBackToMap()

            .assertText(
                getApplicationContext<Context>().resources.getString(
                    R.string.select_item_count,
                    getApplicationContext<Context>().resources.getString(R.string.saved_forms),
                    1,
                    1
                )
            )
            .selectForm(mapFragment, 1)
            .clickEditSavedForm("Single geopoint")
            .clickOnQuestion("Name")
            .assertText("Foo")
    }

    private fun stubGeopointIntent() {
        val location = Location("gps")
        location.latitude = 125.1
        location.longitude = 10.1
        location.altitude = 5.0

        val intent = getReturnIntent(GeoUtils.formatLocationResultString(location))
        val result = Instrumentation.ActivityResult(Activity.RESULT_OK, intent)

        intending(hasComponent("org.odk.collect.geo.geopoint.GeoPointActivity"))
            .respondWith(result)
    }

    companion object {
        private const val SINGLE_GEOPOINT_FORM = "single-geopoint.xml"
        private const val NO_GEOPOINT_FORM = "basic.xml"
    }
}
